### PR TITLE
Add cache support

### PR DESCRIPTION
**Pull request by dsseng**

### CODE REVIEW

File: middleware.js [hunk 1]
```diff
@@ -1,34 +1,108 @@
...
+
+const vueMiddleware = (options = defaultOptions) => {
+  let cache
+  if (options.cache) {
+    const LRU = require('lru-cache')
+
+    cache = new LRU({
+      max: 100,
+      length: function (n, key) { return n * 2 + key.length },
+      maxAge: 1000 * 60 * 60
```
Comment (yyx990803) [inline middleware.js]: I don't think we even need a maxAge here since this is dev only for now.

File: middleware.js [hunk 2]
```diff
@@ -1,34 +1,108 @@
...
+  cache: true
+}
+
+const vueMiddleware = (options = defaultOptions) => {
+  let cache
+  if (options.cache) {
+    const LRU = require('lru-cache')
+
+    cache = new LRU({
+      max: 100,
```
Comment (yyx990803) [inline middleware.js]: Probably should have a bigger limit

File: middleware.js [hunk 3]
```diff
@@ -1,34 +1,108 @@
...
+    return vueCompiler.assemble(compiler, filepath, descriptorResult)
+  }
+
   return async (req, res, next) => {
-    // TODO caching
     if (req.path.endsWith('.vue')) {
-      const { filepath, source } = await readSource(req)
-      const descriptorResult = compiler.compileToDescriptor(filepath, source)
-      const assembledResult = vueCompiler.assemble(compiler, filepath, descriptorResult)
+      const key = parseUrl(req).pathname
```
Comment (yyx990803) [inline middleware.js:64]: The cache for a file needs to be busted whenever the source code changes (user hits save, file written to disk) - simple implementation would be using `fs.stats` to get the last modified timestamp and store that along with the compile result in the cache.

File: middleware.js [hunk 4]
```diff
@@ -1,34 +1,108 @@
...
+        out = (await bundleSFC(req)).code
+        cacheData(key, out)
+        cached = true
+      }
 
-      send(res, assembledResult.code, 'application/javascript')
+      send(res, out, 'application/javascript')
+
+      // Bundle Single-File Component
+      out = (await bundleSFC(req)).code
```
Comment (yyx990803) [inline middleware.js]: not sure why compiling again here

**Review by yyx990803 — CHANGES_REQUESTED** [review]

File: middleware.js (hunk 2 above)
Comment (dsseng) [reply middleware.js]: @yyx990803 resolved!

File: middleware.js [hunk 5]
```diff
@@ -1,34 +1,100 @@
...
+  }
+
   const compiler = vueCompiler.createDefaultCompiler()
 
   function send(res, source, mime) {
     res.setHeader('Content-Type', mime)
     res.end(source)
   }
 
+  async function tryCache (key, upd = true) {
```
Comment (yyx990803) [inline middleware.js]: Can we use a more descriptive name than `upd`?

Comment (yyx990803) [reply middleware.js]: (this should apply for all occurrences)

**Review by yyx990803 — CHANGES_REQUESTED** [review]

File: middleware.js (hunk 5 above)
Comment (dsseng) [reply middleware.js]: @yyx990803 `updateTime`?

Comment (dsseng) [reply middleware.js]: @yyx990803 done

**Review by yyx990803 — APPROVED** [review]

**Pull request merged**